### ISSUE TITLE
Feature Request: cidrsubnets and map Interpolation enhancements

### ISSUE THREAD

**Issue by fstuck37** [hashicorp/terraform#16266]
Hi All,
I've tried to do this with existing interpolation functions like cidrsubnet but can't seem to get the calculations to work 100% of the time. Would love to have a function that could do this to help with automatic subnet allocations based on starting CIDR block.

cidrsubnets: Takes an IP address range in CIDR notation (like 10.0.0.0/8) and a list of prefix sizes and divides the range into a consistent set of subnets.
cidrsubnets( <cidr>, list(<prefix size 1>,...,<prefix size n>))

cidrsubnets( "10.128.0.0/23", ["25","24"])
returns ["10.128.0.0/25", "10.128.1.0/24"]

cidrsubnets( "10.128.0.0/23", ["25","24","25"])
returns ["10.128.0.0/25", "10.128.1.0/24", "10.128.0.128/25"]

cidrsubnets( "10.128.0.0/20", list("22","24","23","24"))
returns ["10.128.4.0/22", "10.128.0.0/24", "10.128.2.0/23", "10.128.1.0/24"]

cidrsubnets( "10.128.0.0/24", list("25","24","25"))
returns an error that subnet list cannot fit in "10.128.0.0/24"

Extend the map function so that your can use a list for both the values and keys such that:
var.map == map(keys(var.map), values(var.map))

Here's an example use case:
variable "net_cidr" {
  default = "10.128.0.0/20"
}

variable "subnets" {
  description = "Subnet sizes of various netorks"
  default = {
    net1 = 22
    net2 = 24
    net3 = 23
    net4 = 24
  }
}

map( keys(var.subnets), cidrsubnets(var.net_cidr, values(var.subnets)) )
resulting map would look like this:
{
    net1 = "10.128.4.0/22"
    net2 = "10.128.0.0/24"
    net3 = "10.128.2.0/23"
    net4 = "10.128.1.0/24"
}

Thanks,
Fred

**Comment by apparentlymart** [comment]
Hi @fstuck37!

This is an interesting idea. It feels like a similar use-case to the CIDR provider proposed in #15106. Although we haven't been able to make that provider "official" yet (there's a bit of a provider-onboarding backlog :confounded:) you may be able to use it by compiling and installing it as a [third-party provider](https://www.terraform.io/docs/configuration/providers.html#third-party-plugins), from this repository: https://github.com/moofish32/terraform-provider-cidr

**Comment by fstuck37** [comment]
Martin,
Looks interesting! Will try to compile it when time permits.
(We'll see who gets to it first 😖)

Thanks for the quick response and for all the work you do on Terraform.
-Fred

**Comment by moofish32** [comment]
@fstuck37 -- If there are features missing in my initial CIDR provider, I would appreciate the feedback. I'm also still working to improve the provider so additional use cases beyond my own are exactly what we need. I started with two simple implementations docs are here:
https://github.com/moofish32/terraform-provider-cidr/blob/master/website/docs/d/network.html.markdown
and
https://github.com/moofish32/terraform-provider-cidr/blob/master/website/docs/d/network.html.markdown
You should be able to pull and use the plugin. If not ping me here and I'll help as much as I can.

**Comment by et304383** [comment]
I would love this.  I'm sorry, but the current cidrsubnet interpolation function is confusing and poorly documented.

**Comment by fstuck37** [comment]
@apparentlymart I did some testing with this and don't see any issues. This looks really good and would accommodate what we need.

Would love to see this move up the on-boarding backlog soon.

Thanks for directing me to this,
Fred

**Comment by fstuck37** [comment]
For everyone's references here is one of my tests and use cases.

/* Sample Code */
variable "cidr_block" { default = "192.168.0.0/19" }
variable "number_azs" { default = "4" }

data "cidr_network" "cake" {
  cidr_block = "${var.cidr_block}"
  subnet {
    mask = 22
    name = "pub"
  }
  subnet {
    mask = 22
    name = "web"
  }
  subnet {
    mask = 21
    name = "app"
  }
  subnet {
    mask = 22
    name = "db"
  }
  subnet {
    mask = 22
    name = "mgt"
  }
}

output "pub" {
  value = "${ data.cidr_network.cake.subnet_cidrs["pub"] }"
}

output "pub1-az1" {
  value = "${ cidrsubnet(data.cidr_network.cake.subnet_cidrs["pub"], log(var.number_azs, 2), 0)  }"
}

output "pub1-az2" {
  value = "${ cidrsubnet(data.cidr_network.cake.subnet_cidrs["pub"], log(var.number_azs, 2), 1)  }"
}

output "pub1-az3" {
  value = "${ cidrsubnet(data.cidr_network.cake.subnet_cidrs["pub"], log(var.number_azs, 2), 2)  }"
}

output "pub1-az4" {
  value = "${ cidrsubnet(data.cidr_network.cake.subnet_cidrs["pub"], log(var.number_azs, 2), 3)  }"
}

/* Sample Output */
pub = 192.168.0.0/22
pub1-az1 = 192.168.0.0/24
pub1-az2 = 192.168.1.0/24
pub1-az3 = 192.168.2.0/24
pub1-az4 = 192.168.3.0/24

**Comment by fstuck37** [comment]
@apparentlymart Just a friendly reminder that some testing was done with this provider. Hoping it can be on-boarded soon. If you require any more testing then what I've completed thus far, please let me know.
Thanks and have a Happy Thanksgiving!,
Fred

**Comment by fstuck37** [comment]
@apparentlymart Just another friendly reminder on this one. I think this provider would be a great addition. Hope to see it added soon.

Also wanted to wish you all a very happy Holiday Season and wonderful New Year.
Thanks for all your hard work and dedication to this project.
-Fred

**Comment by moofish32** [comment]
@apparentlymart -- we've been sitting on this provider for a few months now. Anything I can do to move this forward?

**Comment by apparentlymart** [comment]
Hi @moofish32! Sorry for the long silence here.

We've been reorganizing some teams inside HashiCorp to better support third-party provider development and onboarding. Things haven't totally settled yet, but I think the best path is to submit your provider through [the provider development program](https://www.terraform.io/guides/terraform-provider-development-program.html) to get it into the queue that the new team is working through.

You'll see that the process and form there are oriented towards the more common use-case of an infrastructure/service vendor that wants to submit a provider that wraps a remote API, but there are humans on the other side of that form who should be able to work with you keeping in mind that this is an unusual sort of provider that doesn't interact with any external systems; just fill it out the best you can and we can work out the details afterwards.

**Comment by fstuck37** [comment]
@moofish32 / @apparentlymart 
Any update on on-boarding this provider?
Thanks,
Fred

**Comment by moofish32** [comment]
@fstuck37 -- trying to hit all the threads. As indicated in the PR thread I have not received any information back from the link above.

**Comment by fstuck37** [comment]
@apparentlymart 
Seems like the provider you suggested will not be added unless @moofish32 becomes an official provider per #15106. 
This feature request still stands and the provider moofish32 developed would fulfill this request.
Any suggestions?
Thanks,
Fred

**Comment by crw** [comment]
Thank you for your continued interest in this issue. 

[Terraform version 1.8](https://github.com/hashicorp/terraform/blob/v1.8.0-beta1/CHANGELOG.md) launches with support of provider-defined functions. It is now possible to implement your own functions! We would love to see this implemented as a provider-defined function. 

Please see the [provider-defined functions documentation](https://developer.hashicorp.com/terraform/plugin/framework/functions) to learn how to implement functions in your providers. If you are new to provider development, learn how to [create a new provider](https://developer.hashicorp.com/terraform/plugin/framework) with the Terraform Plugin Framework. If you have any questions, please visit the [Terraform Plugin Development category in our official forum](https://discuss.hashicorp.com/c/terraform-providers/tf-plugin-sdk/43). 

We hope this feature unblocks future function development and provides more flexibility for the Terraform community. Thank you for your continued support of Terraform!

**Comment by apparentlymart** [comment]
Looking back at this issue quite some time later, it _seems_ like it's resolved, but I'm having trouble remembering when this issue was opened in relation to the features that I think might have solved it.

- We added a function [`cidrsubnets`](https://developer.hashicorp.com/terraform/language/functions/cidrsubnets) that seems to match the proposal here.
- The proposed extension to the `map` function (which has since been removed altogether in favor of the built-in object constructor syntax and the `tomap` function)  seems to match [`zipmap`](https://developer.hashicorp.com/terraform/language/functions/zipmap).
- The outcome of the question about the CIDR provided was that we decided to instead add the aforementioned `cidrsubnets` function and then implement the remaining behavior as the module [`hashicorp/subnets/cidr`](https://registry.terraform.io/modules/hashicorp/subnets/cidr/latest), wrapping that function.

I think all of these things taken together addressed the use-cases described in this issue, so I'm going to close it. Sorry for the long delay in doing so!
